### DISCUSSION TITLE
Time Series Experimental Logging with MLFlow

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1885] [FAQs]
# Suggested Previous Reads

   - **[Easy MLOps with PyCaret + MLflow](https://towardsdatascience.com/easy-mlops-with-pycaret-mlflow-7fbcbf1e38c6)** by @moezali1

# Resources for this FAQ

- **[Jupyter Notebook](https://nbviewer.ipython.org/github/ngupta23/pycaret_faqs/blob/main/time_series/pycaret_mlflow_basics.ipynb)** for this FAQ.

# Overview

The `pycaret` timeseries module now supports experiment logging using MLFlow. MLflow is an open-source platform to manage the ML lifecycle, including experimentation, reproducibility, deployment, and a central model registry. You can read more about it in the article mentioned above. In this FAQ, we will focus on how you can log your timeseries experiments using MLFlow.

# Solution

## Basic Logging

Basic logging can be enabled using the following setup arguments: `log_experiment = True` and provide an `experiment_name`.

```python
#### Setup experiment with MLFlow logging ----
exp.setup(
    data=y,
    fh=12,
    session_id=42,
    log_experiment=True,
    experiment_name="my_exp",
  )
```

Once logging is enabled, users can launch the MLFlow server from the notebook directly or from the shell. 

#### From Jupyter Notebook
```python
## Run from terminal that has the python/anaconda environment.
## Alternately, run from notebook as below
## Web browser must be accessible to see results.
# !mlflow ui
```

#### From shell

![image](https://user-images.githubusercontent.com/33585645/143017686-e14c0177-6991-43a7-949e-ff17fd2ca7dc.png)

Then, users can open the server location on your browser to view the logged experiments.

![image](https://user-images.githubusercontent.com/33585645/143015380-1f3aad4b-57eb-49cf-b493-651965321e1b.png)

Let's create a model to see what gets logged in MLFlow. We will create a reduced regression model using Linear Regression. You an read more about Reduced Regression models here: https://github.com/pycaret/pycaret/discussions/1760

```python
#### Create a model ----
model = exp.create_model("lr_cds_dt")
```

This model now shows up in the MLFlow server.

![mlflow1](https://user-images.githubusercontent.com/33585645/143016136-a4999ab8-6278-48bb-a05b-320dd3a5decc.PNG)

Clicking on the model name provides more details such as what hyperparameters were used to create the model. For example, in this case, the model used a `window_length = 10` which is an important hyperparameter for reduced regression models.

![mlflow2](https://user-images.githubusercontent.com/33585645/143016393-eeabaa25-b180-4bb1-9296-8893c8b13855.PNG)

Not only that, MLFlow also shows model artifacts such as the cross-validation metrics and the test (holdout) set results. These can be helpful when reviewing the experiment and models at a later point in time.

#### Cross-Validation Results

![mlflow3](https://user-images.githubusercontent.com/33585645/143016821-20b1dea7-e16a-45ea-9585-94921e96c1c4.PNG)

#### Test (Holdout) Results

![mlflow4](https://user-images.githubusercontent.com/33585645/143016848-db18d13d-a640-4a0b-b38d-64d3d420f6a1.PNG)

## Logging with Plots (Default)

`pycaret` also allows logging of plots in MLFlow. This can be enabled by setting `log_plots = True`

```python
#### Setup experiment with MLFlow logging ----
#### This time, also enable logging of (default) plots ----
exp.setup(
    data=y,
    fh=12,
    session_id=42,
    log_experiment=True,
    experiment_name="my_exp_with_plots",
    log_plots=True
  )
```

By default, this will log the "forecast" plots for estimators. These are logged using `plotly` and are interactive so users can zoom and hover over the plots directly in the MLFlow server to get more information as shown below. Interactive plots can come in very handy when reviewing the results. 

![mlflow6](https://user-images.githubusercontent.com/33585645/143017571-85b529ff-ab7c-4087-92f2-19f55c41cd1d.PNG)

## Logging with Plots (Custom)

Moreover, `pycaret` allows users to customize the plots they would like to log. This can be accomplished by passing a list of plots to be logged to the `log_plots` argument. For example, in this case, we are logging the plots that show

1. What the time series data looks like.
2. How data is split between train and test
3. How the train data is split for cross-validation
4. The forecast made by the model

```python
#### Setup experiment with MLFlow logging ----
#### This time, also enable logging of plots (custom) ----
exp.setup(
    data=y,
    fh=12,
    session_id=42,
    log_experiment=True,
    experiment_name="my_exp_with_custom",
    log_plots=["ts", "train_test_split", "cv", "forecast"]
  )
```

![mlflow7](https://user-images.githubusercontent.com/33585645/143018247-04a09132-145d-4127-b23e-11d0348b40ab.PNG)

# Conclusion

We have seen how easy it is to log experiments in pycaret using MLFlow without writing any additional lines of code. MLFlow also allows for easy organization of the modeling artifacts without any additional coding overhead. This can be powerful when building a handful of models but the power grows exponentially when building forecasting models at scale.

That's it for now. In a future FAQ/article, we will see how we can use pycaret and MLFlow to get a better understanding of the model hyperparameters. Happy forecasting!

# Suggested Next Reads

- [Exploring Model Hyperparameter Search Space with PyCaret & MLflow](https://towardsdatascience.com/exploring-model-hyperparameter-search-space-with-pycaret-mlflow-b6c630d71723)

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
